### DISCUSSION TITLE
What's your opinion on GitHub Action delineation?

### DISCUSSION THREAD

**Discussion by jcbhmr** [spenserblack/spenserblack#45] [General]
For instance, should it be [this (left)](https://github.com/jcbhmr/md-html/blob/1cf318ac430b7b136009a1119c39e80114633d80/.github/workflows/publish.yml) or [this (right)](https://github.com/jcbhmr/md-html/blob/0793ace6a9be3ad4b7b1b9c09d96575176d2abcd/.github/workflows/publish.yml)?

<table><td><img src="https://user-images.githubusercontent.com/61068799/227080796-2f13b4f3-dea1-47e2-bd96-481999ec8af2.png"><td><img src="https://user-images.githubusercontent.com/61068799/227080715-151de162-9906-4cc5-96fb-0fde3c4edde1.png"></table>

There's even a [third idea that i toyed with](https://github.com/jcbhmr/md-html/tree/e5209d084a72ac7028870f17539ada19386db6c3/.github/workflows): separate workflow files for each

![image](https://user-images.githubusercontent.com/61068799/227081340-8cf4f535-7c07-423d-9d83-ffc25b9be11b.png)

(from this project https://github.com/jcbhmr/md-html)

basically where do you draw the line? should you re-`vite build` in each job? how do you avoid 3x re-building for each job? should they be separate files? are actions segmented into files based on "what triggers them" or "what they do" or some other criteria?

i assume from your spenserblack/actions-* projects that you're more experienced with github actions than i am which is why im asking for your opinion 😃

**Comment by spenserblack** [comment]
Good question! I don't think I have strong opinions or hard rules, but here are some things to consider:
- Will significant time be saved by combining into one job? If the build step takes a very long time, you might want to put it in one job. Thankfully, for your examples, Vite is *very* fast.
- If one step fails, should the whole "publish" be considered a failure? This could be a reason to split "publish assets" and "publish docs" into separate jobs. Then you can view/report exactly which part failed easily.
- What do you want to happen when you rerun failed jobs? Again, this could be a reason to split jobs.
- If it's a compiled language, you'll need to consider cross-compiling. For example, Go is *very* easy to cross-compile, so I put that all on one OS (and thus, one job). Other languages are much more of a pain to cross-compile, so I might use a matrix for each OS.

All of the options you posted look good to me, but [I've usually done the second one](https://github.com/imretro/bitio-ts/blob/33e0a6b7c92cbb99391bcf9c01effc6b609a85a8/.github/workflows/release.yml). I think that redundantly calling steps that aren't too resource-consuming is acceptable to break tasks into smaller, more manageable jobs.

**Reply by jcbhmr** [reply]
i guess it just comes down to what _feels_ right. thanks for your input! i too like the second one with separate jobs for each task, i'd just like to be able to trigger a `workflow_dispatch` on _individual jobs_ in that publish.yml file (for debugging -- I made 20 commits fussing with workflows).

**Reply by spenserblack** [reply]
> i'd just like to be able to trigger a `workflow_dispatch` on _individual jobs_

It would be awesome if there were workflow-level triggers, but also job-level triggers.
But, as things are now, wanting to use `workflow_dispatch` is a pretty good reason to split the tasks into separate workflows :slightly_smiling_face:
